source/AI.cpp

```diff
@@ -535,25 +526,44 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 	const int maxMinerCount = minables.empty() ? 0 : 9;
 	bool opportunisticEscorts = !Preferences::Has("Turrets focus fire");
 	bool fightersRetreat = Preferences::Has("Damaged fighters retreat");
+	bool shouldUpdateEscortState = !static_cast<bool>(Random::Int(20));
+	bool playerEscortStateUpdated = false;
 	for(const auto &it : ships)
 	{
 		// Skip any carried fighters or drones that are somehow in the list.
 		if(!it->GetSystem())
 			continue;
 
+		// Update fleet-wide states for NPCs, player ships, and carried ships.
+		if(shouldUpdateEscortState && !Random::Int(10) && (!it->IsYours() || !playerEscortStateUpdated))
+		{
+			if(it->IsYours())
+				playerEscortStateUpdated = true;
+			// Only update escort state if it makes sense to do so.
+			it->UpdateEscortsState();
+		}
```

Comment:
Each time a fleet member is updated it uses the fleet hierarchy.  So it selects the flagship and then updates the fleet state from the flagship down.  If two escorts get selected within the same fleet then the whole fleet is updated twice.  This is more processing than necessary since if run a second time it doesn’t change anything.

Player fleet need only update its state once per frame.  There’s no benefit for two escorts triggering two fleet state updates.

Reply:
Okay now I see the UpdateEscort**s**State() applies on the whole fleet, this means NPC fleets get updated multiple times though? Maybe we only do it if we are dealing with a ship without a parent?

Reply:
If we do it for a ship without a parent I don't know what effect that would have on NPC getting their fleet stats updated.  I'd be willing to try it out, though.

Reply:
Its no big deal, and I'm not sure I grasp the entirety of the logic here, but I feel like there is some redundancy if we just keep it like this, since it updates for all ships in a fleet and not the ship itself.

The logic cutter you used here for the player could be expanded for all fleets, where we just take care of a ship if it is a parent/if it has none? (aka leader of fleet/lone ship)